lib/net.c

```diff
@@ -720,10 +720,62 @@ get_if_back(struct net_config *net_conf)
 	ETH_RSS_NONFRAG_IPV6_OTHER | \
 	ETH_RSS_IPV6_EX)
 
+static int
+check_port_rss_key_update(struct gatekeeper_if *iface, uint16_t port_id,
+	bool *rss_key_updatable, struct rte_eth_rss_conf *rss_conf)
+{
+	struct rte_eth_rss_conf cur_rss_conf;
+	int ret = rte_eth_dev_rss_hash_update(port_id, rss_conf);
+	if (ret == 0)
+		return ret;
+
+	ret = rte_eth_dev_rss_hash_conf_get(port_id, &cur_rss_conf);
+	if (ret < 0) {
+		G_LOG(WARNING,
+			"net: the %s interface does not support a critical hardware capability: update and get the RSS hash configuration\n",
+			iface->name);
+		return ret;
+	}
+
+	/*
+	 * At least two ports with different RSS hash keys
+	 * cannot be updated.
+	 */
+	if (!*rss_key_updatable &&
+			(cur_rss_conf.rss_key == NULL ||
+			cur_rss_conf.rss_key_len !=
+			rss_conf->rss_key_len ||
+			memcmp(rss_conf->rss_key,
+			cur_rss_conf.rss_key,
+			rss_conf->rss_key_len) != 0))
+		return -EINVAL;
+
+	if (cur_rss_conf.rss_key_len > GATEKEEPER_RSS_KEY_LEN)
+		return -EINVAL;
+
+	/* Update the default RSS hash key. */
+	memset(default_rss_key, 0, RTE_DIM(default_rss_key));
+	rte_memcpy(default_rss_key, cur_rss_conf.rss_key,
+		cur_rss_conf.rss_key_len);
+
+	/* Convert RSS key. */
+	rte_convert_rss_key((uint32_t *)&default_rss_key,
+		(uint32_t *)rss_key_be,
+		RTE_DIM(default_rss_key));
```

Comment:
1. `check_port_rss_key_update()` is called on each port and `gatekeeper_init_network()` already calls `rte_convert_rss_key()`. So I think it's fine to drop the call to `rte_convert_rss_key()` here.
2. The call of `rte_convert_rss_key()` was copied from `gatekeeper_init_network()` and it seems to have two issues here:
2.1 Shouldn't we drop `&` from `(uint32_t *)&default_rss_key`?
2.2 `rte_convert_rss_key()` expects the size of `default_rss_key` in bytes, so use `sizeof(default_rss_key)`. Moreover, add `RTE_BUILD_BUG_ON(sizeof(default_rss_key % 4 != 0);` just before the call of `rte_convert_rss_key()` to guarantee that our key has a proper size.

Please add a patch for updating the call of `rte_convert_rss_key()` in `lib/net.c:gatekeeper_init_network()`.

Reply:
2.1 Indeed, the code was initially copied from a [test example](https://github.com/cjdoucette/dpdk/blob/46a945c1a3bf5904cb88c28bf1e8c441e0536641/test/test/test_thash.c#L116-L117) from DPDK. I just found another [example](https://github.com/cjdoucette/dpdk/blob/46a945c1a3bf5904cb88c28bf1e8c441e0536641/lib/librte_eventdev/rte_event_eth_rx_adapter.c#L2003-L2005), which removes ``&`` from ``(uint32_t *)&default_rss_key``. By checking the [definition](https://github.com/cjdoucette/dpdk/blob/46a945c1a3bf5904cb88c28bf1e8c441e0536641/lib/librte_hash/rte_thash.h#L153-L160) of ``rte_convert_rss_key()``, I think you are right.

Reply:
I have run the [test example](https://github.com/cjdoucette/dpdk/blob/46a945c1a3bf5904cb88c28bf1e8c441e0536641/test/test/test_thash.c#L116-L117) with and without ``&``, both pass the tests. Here is an [explanation](https://stackoverflow.com/questions/8412694/address-of-array).

Reply:
The link to the explanation was very useful. Thank you.

Let's have it without the `&`. The other form is unusual.